### ISSUE TITLE
Criterion, Gauge, Tasty-bench chapter

### ISSUE THREAD

**Issue by doyougnu** [input-output-hk/hs-opt-handbook.github.io#33]
combine these because of similar api

**Comment by Bodigrim** [comment]
The basic API is roughly similar, but capabilities provided are very different. I could probably contribute a chapter on `tasty-bench` usage, if this is of any interest.

**Comment by doyougnu** [comment]
Hi @Bodigrim. Yes that would be great but pleas wait for a little while. David Christiansen and I are in the process of moving this repository to the HF github and so I want to keep it as simple as possible until that happens. Should be within the next two weeks.

**Comment by doyougnu** [comment]
@Bodigrim several months later the migration is stuck trying to pass some of IOGs internal checks. So I think its better to not wait, would you still like to contribute the chapter?

**Comment by Bodigrim** [comment]
If your plan is to describe several benchmarking libraries, we need to agree on the structure. 

I can submit Chapter 2.5.1 on `tasty-bench` and then further potential contributors may elaborate about `criterion` / `gauge` differences in a hypothetical Chapter 2.5.2. Or we can do it vice versa, but in such case I obviously would not be able to start before a description of `criterion` / `gauge` is done. I'm in favor of the former option, are you comfortable with it?

Besides that, I'm not a native English speaker. Would you be happy to proof-read?

**Comment by doyougnu** [comment]
> I can submit Chapter 2.5.1 on tasty-bench and then further potential contributors may elaborate about criterion / gauge differences in a hypothetical Chapter 2.5.2.

Yes this is a good plan! What I had in mind for the chapter was not to repeat the tutorial for the library but to instead describe a real use case. So for example, start with the library that you'll use tasty-bench to benchmark, describe a bit about the library (like where do we expect the heavy computation to be, which code should we even be benchmarking?), and then describe how the benchmark suite was implemented using tasty-bench. The reason I envision the chapter this way is because it starts where the audience is, i.e, they are looking to use tasty-bench to benchmark something and want a real-world example, not just benchmarking fib. 

> Besides that, I'm not a native English speaker. Would you be happy to proof-read?

Absolutely! :)

**Comment by Bodigrim** [comment]
> What I had in mind for the chapter was not to repeat the tutorial for the library but to instead describe a real use case. So for example, start with the library that you'll use tasty-bench to benchmark, describe a bit about the library (like where do we expect the heavy computation to be, which code should we even be benchmarking?), and then describe how the benchmark suite was implemented using tasty-bench.

Fair enough, but this is different from the current presentation of `weigh`, which is more a tutorial than a case study.

**Comment by doyougnu** [comment]
> Fair enough, but this is different from the current presentation of weigh, which is more a tutorial than a case study.

Ah very true. Let's do a tutorial then. I think it is better to be consistent than to do something special for tasty-bench. My issue with the `weigh` chapter was that I felt that it is a useful library _without_ a good tutorial. I'm not entirely sure I got to _good_ tutorial but I do feel that it improved the documentation around weigh. So I suppose I'm being a bit biased towards `tasty-bench` because I think the `tasty` suite of packages are already very well documented, they are just documented using micro-examples. What I'm looking for is more "how to use tasty-bench to benchmark my _system_" and less "how to use tasty-bench to benchmark this or that function". Hopefully that makes sense.

**Comment by Bodigrim** [comment]
I actually like the idea of writing case studies instead of tutorials; I think the official README for `tasty-bench` is reasonably good and there is probably not much point to retell it. The question is more about whether you'll be able to integrate case studies in the fabric and structure of the book.

**Comment by Bodigrim** [comment]
> The key for me is that the chapter has the perspective of a system, not just a code snippet.

This might be challenging in the context of a book of limited volume. But I see your point. 

I think my best option is to write some blog posts on benchmarking and dual license them so that you can potentially include them into the book and edit as necessarily.
